ruby/computer_science/time_complexity.md

```diff
@@ -0,0 +1,306 @@
+* What to do when two algorithms have the same complexity.
+
+### Efficiency Basics
+
+The very first step in mastering efficient code is to understand how to measure it. Let's take a look at a simple little programme that prints out all odd numbers between 1 and 10.
+
+~~~ruby
+def odd_numbers_less_than_ten
+  current_number = 1
+
+  while current_number < 10
+    if current_number % 2 != 0
+      puts current_number
+    end
+
+    current_number += 1
+  end
+end
+~~~
+
+If you were to run this in your terminal you should get the numbers `1`, `3`, `5`, `7` and `9` printed to the console. It probably took a fraction of a second to run. If you were to run it again it might take the same time, or it might be faster or slower depending on what else your computer is doing. If you were to run it on a different computer, it would again run faster or slower. Therefore it's important to understand that you never measure the efficiency of an algorithm by how long it takes to execute.
+
+So how do we measure it?
+
+The way to measure code efficiency is to evaluate how many 'steps' it takes to complete. If you know that one algorithm you write takes 5 steps and another one takes 20 steps to accomplish the same task then you can say that the 5 step algorithm will always run faster than the 20 step algorithm on the same computer.
+
+Let's go back to our `odd_numbers_less_than_ten method`. How many steps does our algorithm take?
+
+1. We assign the number 1 to a variable. That's one step.
+
+2. We have a loop. For each iteration of the loop we do the following.
+    1. Compare `current_number` to see if it is less than 10, that is 1 step.
+    2. We then check if current_number is odd, that is 1 step.
+    3. If it is then we output it to the terminal. That's 1 step every 2 iterations.
+    4. We increase `current_number` by 1. That is 1 step.
+
+So there are 3 steps for every loop iteration and it iterates 9 times which is 27 steps. Then we have one step which iterates for only half the loop iteration which is 5 steps. Assigning an initial value to `current_number` is just one step. 27 + 5 + 1 = 33 steps.
+
+Therefore we can say our algorithm takes 33 steps to complete.
+
+While this is useful to know it isn't actually helpful for comparing algorithms. To see why let's slightly modify our initial algorithm to take in a number instead of set a hard default of 10.
+
+~~~ruby
+def odd_numbers(max_number)
+  current_number = 1
+
+  while current_number < max_number
+    if current_number % 2 != 0
+      puts current_number
+    end
+
+    current_number += 1
+  end
+end
+~~~
+
+How many steps does this algorithm take?
+
+You've probably realised the answer is it depends. If you enter 10 like we did before the number of steps is 33 but if you enter another number then the number of steps changes. There is no concrete number we can use to measure the efficiency of our code because it changes based on an external input.
```

Comment:
We could be more explicit here: 
```suggestion
You've probably realised the answer is it depends. If you set `max_number` to be 10 like we did before the number of steps is 33 but if you enter another number then the number of steps changes. There is no concrete number we can use to measure the efficiency of our code because it changes based on an external input.
```